File: extension/src/broker/message/log.ts
import { Squash, type SerDes } from "squash-ts";
import { serdeArray, str } from "./serde";

/**
 * The severity of one relayed console line.
 */
export type LogLevel = "print" | "info" | "warn" | "error";

/** One line of a client's console output. */
export interface LogEntry {
    level: LogLevel;
    content: string;
}

/** Mirrors `LogLevel`. */
export const logLevel: SerDes<LogLevel> = Squash.literal<LogLevel>("print", "info", "warn", "error");

/** A plain struct, so `ser` runs forward and `des` in reverse. */
export const logEntry: SerDes<LogEntry> = {
    ser(cursor, entry) {
        logLevel.ser(cursor, entry.level);
        str.ser(cursor, entry.content);
    },

    des(cursor) {
        const content = str.des(cursor);
        return { level: logLevel.des(cursor), content };
    },
};

/** Mirrors `Vec<LogEntry>`. */
export const logEntries = serdeArray(logEntry);

File: broker/src/protocol/log.rs
use serde::{Deserialize, Serialize};
use squash::ReverseDeserialize;

/// The severity of one relayed console line.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum LogLevel {
    Print,
    Info,
    Warn,
    Error,
}

/// One line of a client's console output.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, ReverseDeserialize)]
pub struct LogEntry {
    pub level: LogLevel,
    pub content: String,
}
